wagtail/models/__init__.py

```diff
@@ -2907,20 +2908,123 @@ def __str__(self):
+                )
+                .prefetch_related(prefetch_user_perms)
+                .distinct()
+                .in_bulk()
+            )
+            # Since a lot of queries involve looking up pages
+            # the user has a specific permission for, we
+            # build this mapping of `permission_type: {page_pks}`
+            # to retrieve that information quickly without asking
+            # the database.
+            self.perm_types = defaultdict(set)
+            for page_pk, page in self._pages.items():
+                # Rewrite the `_perms_for_user` annotation and only
+                # store the different type of permission the user has.
+                page._perms_for_user = {
+                    perm.permission_type for perm in page._perms_for_user
+                }
+
+                for perm_type in page._perms_for_user:
+                    self.perm_types[perm_type].add(page_pk)
+
+        return user._page_permissions_proxy
+
+    def get_pages_for_perms(self, perm_types):
+        """
+        Returns an iterator over all the pages the user
+        has permissions matching `perm_types`.
+        """
+        page_pks = set()
+        for perm_type in perm_types:
+            pages_for_perm = self.perm_types.get(perm_type, None)
+            if pages_for_perm is not None:
+                page_pks |= pages_for_perm
+
+        return (self._pages[page_pk] for page_pk in page_pks)
+
+    @property
+    def pages(self):
+        """
+        Returns an iterator over all the pages the user has any permission for.
+        """
+        return self._pages.values()
+
+    def has_any_page_permission(self):
+        """Indicates whether the user has any permission on any page."""
+        return bool(self._pages)
+
+    def for_page(self, page):
+        """
+        Return a PagePermissionTester object that can be used to query whether
+        this user has permission to perform specific tasks on the given page.
+        """
+        return PagePermissionTester(self, page)
 
+    def pages_with_direct_explore_permission(self):
+        """
+        Get all pages that the user has direct add/edit/publish/lock permission on.
+        """
+        if self.user.is_superuser:
+            # superuser has implicit permission on the root node
+            return Page.objects.filter(depth=1)
+        else:
+            return self.get_pages_for_perms(["add", "edit", "publish", "lock"])
+
+    @cached_property
+    def _explorable_root_page(self):
+        """
+        Get the highest common explorable ancestor for the given user.
+        If the user has no permissions over any pages, this method will return None.
+        """
+        pages = self.pages_with_direct_explore_permission()
+        try:
+            root_page = first_common_ancestor(pages, include_self=True, strict=True)
+        except Page.DoesNotExist:
+            root_page = None
+
+        return root_page
```

Comment:
Nit: You can avoid unnecessary assignment here

```suggestion
            return first_common_ancestor(pages, include_self=True, strict=True)
        except Page.DoesNotExist:
            return None
```